### DISCUSSION TITLE
Making template creation more user friendly

### DISCUSSION THREAD

**Discussion by Conor-Behard333** [coder/coder#12558] [RFCs]
Hi, I'm part of a platforms team and we've been working on integrating coder into our workflow and the workflow of other developers in the company. Across the company we have many different dev needs. Some teams needing python, C, JavaScript, java others needing helm, flux, kubectl, etc. It ends up needing a lot of templates for specific needs. However, Not everyone is familiar with terraform and so the responsibility of creating templates comes down to only a few people (platforms). 

I recently thought of a solution to this; adding a more user friendly (users who lack terraform knowledge) way of creating templates. 

Assuming a terraform module registry is linked to coder, a user could select the terraform modules they want to add into a new template in a "drag and drop" style UI.

Workflow
A user can create a template, using a custom base template. They can then select the modules from the given TF registry. For example, the user selects a base template to create the workspace on a k8s cluster. The user then selects the modules they want (say one module to install java and another to install python). Once the user has selected all the desired modules it can be saved as a full template and used like any normal template.  

From a technical perspective it would just be a UI that can collate modules from a registry which can then form a complete terraform template and treat them more like objects that users can interact with. Rather than dealing directly with terraform code. Making these modules into essentially objects that users can interact with would also allow for permissions on what users are allowed to add into their template.

I'm aware that something like this would not be a simple creation and may be a case of overengineering a problem that maybe only we are having. I'm simply intrigued on what other people thought about it and if it is even feasible.

**Comment by matifali** [comment]
> I recently thought of a solution to this; adding a more user friendly (users who lack terraform knowledge) way of creating templates.

> Assuming a terraform module registry is linked to coder, a user could select the terraform modules they want to add into a new template in a "drag and drop" style UI.

@kylecarbs @bpmct I remember we discussed this once as part of #10664.
